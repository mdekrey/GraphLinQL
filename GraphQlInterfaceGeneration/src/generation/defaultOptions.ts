import { Options } from "./Options";

export const defaultOptions: Options = {
  namespace: "GraphLinqQL.Interfaces",
  useNullabilityIndicator: true,
  scalarTypes: {
    Int: { csharpNullable: false, csharpType: "int" },
    Float: { csharpNullable: false, csharpType: "double" },
    String: { csharpNullable: true, csharpType: "string" },
    Boolean: { csharpNullable: false, csharpType: "bool" }
  },
  using: [
    "GraphLinqQL",
    "GraphLinqQL.Introspection",
    "System",
    "System.Collections",
    "System.Collections.Generic",
    "System.Collections.Immutable",
    "System.Linq"
  ],
  deserializer: "System.Text.Json.JsonSerializer.Deserialize",
  introspection: false
};
