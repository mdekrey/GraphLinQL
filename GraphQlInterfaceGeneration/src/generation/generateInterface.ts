import { GraphQLField, GraphQLArgument, GraphQLInterfaceType } from "graphql";
import { Options } from "./Options";
import { getTypeName } from "./getTypeName";
import { getPropertyName } from "./getPropertyName";
import { getFieldName } from "./getFieldName";
import { getOutputTypeName } from "./getOutputTypeName";
import { getInputTypeName } from "./getInputTypeName";

export function generateInterface(object: GraphQLInterfaceType, options: Options) {
  const typeName = getTypeName(object.name, options);

  const fields = object.getFields();
  return `${
    object.description
      ? `
/// <summary>
/// ${object.description.split("\n").join("\n/// ")}
/// </summary>`
      : `
`
  }
public interface ${typeName} : IGraphQlResolvable
{
    ${Object.keys(fields)
      .map(fieldName => fields[fieldName])
      .map(field => resultAbstractDeclaration(field, options)).join(`
    `)}
}
`;
}

function resultAbstractDeclaration(field: GraphQLField<any, any, { [key: string]: any }>, options: Options) {
  const propertyName = getPropertyName(field.name, options);
  const typeName = getOutputTypeName(field.type, options);
  const args = field.args.map(arg => resultAbstractDeclarationArg(arg, options)).join(", ");
  return `${
    field.description
      ? `
      /// <summary>
      /// ${field.description.split("\n").join("\n    /// ")}
      /// </summary>
      `
      : ``
  }IGraphQlResult${typeName && `<${typeName}>`} ${propertyName}(${args});`;
}

function resultAbstractDeclarationArg(arg: GraphQLArgument, options: Options) {
  const fieldName = getFieldName(arg.name, options);
  const inputTypeName = getInputTypeName(arg.type, options);
  return `${inputTypeName} ${fieldName}`;
}
