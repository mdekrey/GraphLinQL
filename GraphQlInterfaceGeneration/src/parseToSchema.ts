import { parse, buildASTSchema, GraphQLSchema } from "graphql";

export function parseToSchema(graphQlSchema: string): GraphQLSchema {
  const astDocument = parse(graphQlSchema);

  const backcompatOptions = { commentDescriptions: true };
  const schema = buildASTSchema(astDocument, backcompatOptions);

  return schema;
}
