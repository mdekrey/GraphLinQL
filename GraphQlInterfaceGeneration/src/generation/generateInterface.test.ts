import { parseToSchema } from "../parseToSchema";
import { getStarWarsGraphQlSchemaV4, getStarWarsGraphQlSchemaV3 } from "../resources.test";
import { GraphQLInterfaceType } from "graphql";
import { defaultOptions } from "./defaultOptions";
import { generateInterface } from "./generateInterface";

it("can generate simple interfaces", async function() {
  const schema = await getSampleSchema(getStarWarsGraphQlSchemaV4);
  const type = schema.getType("Node") as GraphQLInterfaceType;
  expect(generateInterface(type, defaultOptions)).toMatchSnapshot();
});

it("can generate simple interfaces from v3", async function() {
  const schema = await getSampleSchema(getStarWarsGraphQlSchemaV3);
  const type = schema.getType("Character") as GraphQLInterfaceType;
  expect(generateInterface(type, defaultOptions)).toMatchSnapshot();
});

async function getSampleSchema(getSchema: () => Promise<string>) {
  const schema = parseToSchema(await getSchema());
  return schema;
}
