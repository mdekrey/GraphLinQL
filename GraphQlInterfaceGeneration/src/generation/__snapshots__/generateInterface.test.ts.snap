// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can generate simple interfaces 1`] = `
"
/// <summary>
/// An object with an ID
/// </summary>
public interface Node
{
    
      /// <summary>
      /// The id of the object.
      /// </summary>
      IGraphQlResult<string> id();
}
"
`;

exports[`can generate simple interfaces from v3 1`] = `
"
/// <summary>
/// A character from the Star Wars universe
/// </summary>
public interface Character
{
    
      /// <summary>
      /// The ID of the character
      /// </summary>
      IGraphQlResult<string> id();
    
      /// <summary>
      /// The name of the character
      /// </summary>
      IGraphQlResult<string> name();
    
      /// <summary>
      /// The friends of the character, or an empty list if they have none
      /// </summary>
      IGraphQlResult friends();
    
      /// <summary>
      /// The friends of the character exposed as a connection with edges
      /// </summary>
      IGraphQlResult<FriendsConnection> friendsConnection(int? first, string? after);
    
      /// <summary>
      /// The movies this character appears in
      /// </summary>
      IGraphQlResult<IEnumerable<Episode?>> appearsIn();
}
"
`;
